packages/compass-connections/src/connections-manager.ts

```diff
@@ -122,31 +144,76 @@ export class ConnectionsManager extends EventEmitter {
    * @param connectionInfo The adjusted ConnectionInfo object that already has
    * parameters such as appName.
    */
-  async connect(connectionInfo: ConnectionInfo): Promise<DataService> {
+  async connect(
+    originalConnectionInfo: ConnectionInfo,
+    {
+      appName,
+      forceConnectionOptions,
+      browserCommandForOIDCAuth,
+      onNotifyOIDCDeviceFlow,
+      onDatabaseSecretsChange,
+    }: {
+      appName?: string;
+      forceConnectionOptions: [key: string, value: string][];
+      browserCommandForOIDCAuth?: string;
+      onNotifyOIDCDeviceFlow: OnNotifyOIDCDeviceFlow;
+      onDatabaseSecretsChange: OnDatabaseSecretsChangedCallback;
+    }
+  ): Promise<DataService> {
     try {
       const existingDataService = this.getDataServiceForConnection(
-        connectionInfo.id
+        originalConnectionInfo.id
       );
+
       if (
         existingDataService &&
-        this.statusOf(connectionInfo.id) === ConnectionStatus.Connected
+        this.statusOf(originalConnectionInfo.id) === ConnectionStatus.Connected
       ) {
         return existingDataService;
       }
 
+      const connectionInfo = merge(
+        cloneDeep(originalConnectionInfo),
+        this.oidcState.get(originalConnectionInfo.id) ?? {}
+      ) as ConnectionInfo;
+
+      const isOIDCConnectionAttempt = isOIDCAuth(
+        connectionInfo.connectionOptions.connectionString
+      );
+
+      const adjustedConnectionInfoForConnection: ConnectionInfo = merge(
+        cloneDeep(connectionInfo),
+        {
+          connectionOptions: adjustConnectionOptionsBeforeConnect({
+            connectionOptions: connectionInfo.connectionOptions,
+            defaultAppName: appName,
+            preferences: { forceConnectionOptions, browserCommandForOIDCAuth },
+            notifyDeviceFlow: !isOIDCConnectionAttempt
+              ? undefined
+              : onNotifyOIDCDeviceFlow ??
+                (() => {
+                  return;
+                }),
+          }),
+        }
+      );
+
       this.updateAndNotifyConnectionStatus(
-        connectionInfo.id,
+        adjustedConnectionInfoForConnection.id,
```

Comment:
Is it possible for id to get changed? Are we expecting this for some reason? Feels off that we are using the "adjusted" id here instead of whatever was passed initially. If we don't care about id when modifying the options, can we destructure first in the arguments? Then you can keep the id unchanged and pass the actual options around when modifying them

Reply:
The id is not changing in theory (no one is disallowing it, I suppose a function could broken but we should realise earlier).

I'm not sure I really understand your suggestion. Are you suggesting this:

```js
const { id, ...adjustedConnectionInfo } = merge(...)
```

Or destructuring in the updateAndNotifyConnectionStatus method signature?

Reply:
At the point of the merge it's already modified with whatever is stored in the manager state though, right? I was thinking maybe at the very beginning of the connect function if it's doing all the modifications:

```
async connect({ id: connectionId, ...originalConnectionInfo }) {
  ...
```

It would be nice if connection id would not be so intertwined with the connection options that we actually care about here for modification and passing down to actually connect, but for now I'd at least try to find a way to make it clear in the code that we're always refer to the same id and it can't be changed by anything